File: rust/src/benchmarks/knuckeotide.rs
use super::super::{helper, Benchmark};
use crate::benchmarks::fasta::Fasta;
use crate::config_i64;
use std::collections::HashMap;

pub struct Knuckeotide {
    seq: String,
    result_str: String,
    n: i64,
}

impl Knuckeotide {
    pub fn new() -> Self {
        let n = config_i64("CLBG::Knuckeotide", "n");

        Self {
            n,
            seq: String::new(),
            result_str: String::new(),
        }
    }

    fn frequency<'a>(&self, seq: &'a str, length: usize) -> (usize, HashMap<String, i32>) {
        let n = seq.len().saturating_sub(length) + 1;
        let mut table = HashMap::new();

        for i in 0..n {
            let slice = &seq[i..i + length];
            let key = slice.to_string();
            *table.entry(key).or_insert(0) += 1;
        }

        (n, table)
    }

    fn sort_by_freq(&mut self, seq: &str, length: usize) {
        let (n, table) = self.frequency(seq, length);
        let mut entries: Vec<_> = table.into_iter().collect();

        entries.sort_by(|a, b| b.1.cmp(&a.1).then(a.0.cmp(&b.0)));

        for (seq_str, count) in entries {
            let freq = (count as f64 * 100.0) / n as f64;
            self.result_str
                .push_str(&format!("{} {:.3}\n", seq_str.to_uppercase(), freq));
        }
        self.result_str.push('\n');
    }

    fn find_seq(&mut self, seq: &str, pattern: &str) {
        let (_n, table) = self.frequency(seq, pattern.len());
        let count = table.get(&pattern.to_lowercase()).copied().unwrap_or(0);
        self.result_str
            .push_str(&format!("{}\t{}\n", count, pattern.to_uppercase()));
    }
}

impl Benchmark for Knuckeotide {
    fn name(&self) -> String {
        "CLBG::Knuckeotide".to_string()
    }

    fn prepare(&mut self) {
        let mut fasta = Fasta::new();
        fasta.n = self.n;
        fasta.run(0);

        let result_str = fasta.get_result();

        let mut three_section = false;
        let mut seq_buf = String::new();

        for line in result_str.lines() {
            if line.starts_with(">THREE") {
                three_section = true;
                continue;
            }
            if line.starts_with('>') && three_section {
                break;
            }
            if three_section {
                seq_buf.push_str(line.trim());
            }
        }

        self.seq = seq_buf;
    }

    fn run(&mut self, _iteration_id: i64) {
        let seq_clone = self.seq.clone();

        for i in 1..=2 {
            self.sort_by_freq(&seq_clone, i);
        }

        let patterns = [
            "ggt",
            "ggta",
            "ggtatt",
            "ggtattttaatt",
            "ggtattttaatttatagt",
        ];
        for pattern in patterns {
            self.find_seq(&seq_clone, pattern);
        }
    }

    fn checksum(&self) -> u32 {
        helper::checksum_str(&self.result_str)
    }
}

File: kotlin/src/main/kotlin/benchmarks/Knuckeotide.kt
package benchmarks

import Benchmark

class Knuckeotide : Benchmark() {
    private lateinit var seq: String
    private lateinit var output: StringBuilder

    override fun prepare() {
        output = StringBuilder()

        val fasta = Fasta()
        fasta.n = configVal("n")
        fasta.prepare()
        fasta.run(0)
        val res = fasta.getOutput()

        var three = false
        val seqio = StringBuilder()

        res.lineSequence().forEach { line ->
            if (line.startsWith(">THREE")) {
                three = true
                return@forEach
            }
            if (three) {
                seqio.append(line.trim())
            }
        }

        seq = seqio.toString()
    }

    private fun frequency(
        seq: String,
        length: Int,
    ): Pair<Int, Map<String, Int>> {
        val n = seq.length - length + 1
        val table = mutableMapOf<String, Int>()

        for (f in 0 until n) {
            val sub = seq.substring(f, f + length)
            table[sub] = table.getOrDefault(sub, 0) + 1
        }

        return Pair(n, table)
    }

    private fun sortByFreq(
        seq: String,
        length: Int,
    ) {
        val (n, table) = frequency(seq, length)

        table
            .toList()
            .sortedWith(
                compareByDescending<Pair<String, Int>> { it.second }
                    .thenBy { it.first },
            ).forEach { (key, value) ->
                val freq = (value * 100).toDouble() / n
                output.append(String.format("%s %.3f\n", key.uppercase(), freq))
            }

        output.append('\n')
    }

    private fun findSeq(
        seq: String,
        s: String,
    ) {
        val (n, table) = frequency(seq, s.length)
        output.append("${table.getOrDefault(s, 0)}\t${s.uppercase()}\n")
    }

    override fun run(iterationId: Int) {
        for (i in 1..2) {
            sortByFreq(seq, i)
        }

        listOf("ggt", "ggta", "ggtatt", "ggtattttaatt", "ggtattttaatttatagt").forEach { s ->
            findSeq(seq, s)
        }
    }

    override fun checksum(): UInt = Helper.checksum(output.toString())

    override fun name(): String = "CLBG::Knuckeotide"
}
